aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+// TODO:  Remove this dependency once callBoxed is possible for all ops
+#include <torch/csrc/jit/operator.h>
+
+#include <iostream>
+#include <exception>
+
+namespace at {
+namespace autocast {
+
+namespace {
+// Imitate Apex and cache some of the casts to streamline parameter reuse.
+// Our heuristic is to cache FP16 casts of FP32 model weights (see cached_cast below).
+//
+// After discussion with @ezyang, the cache uses the following structure:
+// The key is the source tensor's TensorImpl*, a proxy for a Tensor uuid that's unchanged across shallow copies.
+// The value is a tuple of two Tensors:  the source tensor and the casted tensor.  Both are stored as full
+// Tensors to ensure their TensorImpls stay alive if other references to them die.
+//
+// We must keep the source tensor alive because if the source tensor were deallocated, another random Tensor
+// could be allocated whose TensorImpl* happened to have the same value.  This TensorImpl* would then mistakenly
+// hit in cache, which would be a nasty bug (rare, intermittent, unpredictable).
+//
+// Since our heuristic caches casts for model weights only, the source Tensors should always stay alive anyway,
+// because the model stores them explicitly.  So in the common case, storing a live reference to the
+// source tensor in cached_casts is unnecessary but also does no harm (does not increase memory use).
+//
+// When the autocast context manager exits, which should occur at the end of each forward pass, it calls
+// clear_cache to ensure cached Tensors don't leak outside the autocasting region.
+thread_local std::unordered_map<TensorImpl*, std::tuple<Tensor, Tensor>> cached_casts;
+}
+
+void clear_cache() {
+  cached_casts.clear();
+}
+
+enum class CastPolicy : uint8_t {
```

Comment:
After reading the enum entries, I'm still not sure what a cast policy is. I imagine I'll figure it out as I continue reading the patch, but it sure would be nice if there were a comment telling me what it is right here ;)

Reply:
Answer: it drives the metaprogramming generation

Reply:
Added comment